game-core/src/main/java/games/strategy/engine/history/change/HistoryChange.java

```diff
@@ -0,0 +1,8 @@
+package games.strategy.engine.history.change;
+
+import games.strategy.engine.delegate.IDelegateBridge;
+
+public interface HistoryChange {
```

Comment:
A quick javadoc could perhaps be useful to explain what constitutes a history change.

I wonder though, wouldn't almost any change be eligible for the history? Is there a (ideally) a 1:1 correspondence between history change objects and game state changes? Are we going to have "peer" change objects in this manner?

Reply:
When you say "game state changes", I'm assuming you are talking about the `Change` objects.

There will not be a 1:1 correspondence between history change objects and game state changes.  A history change object can comprise of 1 or more game state changes.  Take for example `TransformDamagedUnits`.  It creates at minimum an `AddUnits` and a `RemoveUnits` change object.  It also can create many more depending on if attribute translation is asked for and depending on which attributes need to be translated.

I see the `HistoryChange` as more of a higher level object than the `Change` object.  It maps to what the user sees and to what is listed in the history.  The `Change` object is the low level changes that must be done to keep the game state in sync.

One possible benefit of the `HistoryChange` is that if it was stored in the game data history, then it would actually be possible to undo each history entry because the changes are truly grouped together per history entry.

I do plan to add more `HistoryChange` objects.

Reply:
Since they are related, I wonder if instead there is a way to map the history changes from game state changes. Otherwise it seems we are duplicating information about changes to the game state via another change object mechanism.

Reply:
I wouldn't say this is "duplicating information".  I'd say this is grouping information and actually mapping history changes with their underlying game state changes.  Because a history change (and I'm not talking about the history change objects in this PR) might incorporate multiple game state changes.

Reply:
A relationship of many:1 for game state to history changes is interesting. I wonder if that relationship could also be viewed as [0|1] : 1

The reason I am focusing on this is to see if we can simplify how the history is rendered and displayed relative to game state. If we have both a stack of history objects, and a stack of game state objects, how do we easily "point" the game state to an arbitrary location in that game state stack and easily then recreate the history? Or from another point of view, what if the current game is "pointed" to an arbitrary location in history, how do we easily then unwind the game state to that location?

If there is a link (likely bidirectional)  between history objects and state changes, this I think becomes much easier. If such a link is bidirectional, is it potentially actually part of the same data entity? Would it make sense to even just have an API on game state changes like `Optional<HistoryChange> getHistoryElement`

Reply:
As I mentioned, if `HistoryChange` was stored in the history, then it could keep track of all of the `Change`s that occur during that history.  And then if you revert a `HistoryChange` (which this PR doesn't include), it would revert all of the `Change` objects that were part of the history event.

The current mechanism that tracks history items with change objects is done in `games.strategy.engine.history.History`.  I believe how that works is that at some point in the game flow, a call to `HistoryWriter.startEvent` will be called.  This adds a new node to `History`.  When `bridge.addChanges` is called, the changes are stored in a big list inside of `History`.  The nodes created by `startEvent` store the index in the changes list when they are added.  When you traverse the history, it looks up the index of the history node, computes a delta of changes from the big list of changes and then undoes those changes.

That logic feels really convoluted.  I'm not entirely sure I understand it completely.  These `HistoryChanges`, though, could simplify that.  Instead of storing indices into a list of changes and assigning them to each node, the history node would actually store the list of changes that it performed.  And each change object would be attached to a history node, where as now, it looks like you can have change objects that aren't attached to history nodes.